dill/session.py

```diff
@@ -0,0 +1,220 @@
+        if any(obj is singleton for singleton in (None, False, True)) or \
+                isinstance(obj, ModuleType) and _is_builtin_module(obj):  # always saved by ref
+            original[name] = obj
+            continue
+
+        source_module, objname = _lookup_module(modmap, name, obj, main_module)
+        if source_module:
+            if objname == name:
+                imported.append((source_module, name))
+            else:
+                imported_as.append((source_module, objname, name))
+        else:
+            try:
+                imported_top_level.append((modmap.top_level[id(obj)], name))
+            except KeyError:
+                original[name] = obj
+
+    if len(original) < len(main_module.__dict__):
+        newmod.__dict__.update(original)
+        newmod.__dill_imported = imported
+        newmod.__dill_imported_as = imported_as
+        newmod.__dill_imported_top_level = imported_top_level
+        return newmod
+    else:
+        return main_module
+
+def _restore_modules(unpickler, main_module):
+    try:
+        for modname, name in main_module.__dict__.pop('__dill_imported'):
+            main_module.__dict__[name] = unpickler.find_class(modname, name)
+        for modname, objname, name in main_module.__dict__.pop('__dill_imported_as'):
+            main_module.__dict__[name] = unpickler.find_class(modname, objname)
+        for modname, name in main_module.__dict__.pop('__dill_imported_top_level'):
+            main_module.__dict__[name] = __import__(modname)
+    except KeyError:
+        pass
+
+def _filter_objects(main, exclude_extra, include_extra, obj=None):
+    filters = ExcludeRules(getattr(settings, 'session_exclude', None))
+    if exclude_extra is not None:
+        filters.update([(EXCLUDE, exclude_extra)])
+    if include_extra is not None:
+        filters.update([(INCLUDE, include_extra)])
+
+    namespace = filters.filter_namespace(vars(main), obj=obj)
+    if namespace is vars(main):
+        return main
+
+    main = ModuleType(main.__name__)
+    vars(main).update(namespace)
+    return main
+
+def dump_session(filename: Union[PathLike, BytesIO] = '/tmp/session.pkl',
+                 main: Union[str, ModuleType] = '__main__',
+                 byref: bool = False,
+                 exclude: Union[Filter, Iterable[Filter]] = None,
+                 include: Union[Filter, Iterable[Filter]] = None,
+                 **kwds) -> NoReturn:
+    """pickle the current state of __main__ to a file"""
+    protocol = settings.protocol
+    if isinstance(main, str):
+        main = _import_module(main)
+    original_main = main
+    if byref:
+        #NOTE: *must* run before _filter_objects()
+        main = _stash_modules(main)
+    main = _filter_objects(main, exclude, include, obj=original_main)
+    if hasattr(filename, 'write'):
+        f = filename
+    else:
+        f = open(filename, 'wb')
+    try:
+        pickler = Pickler(f, protocol, **kwds)
```

Comment:
It looks like you've done some cleanup of the use of dynamic attributes.  It looks like at least `_file` and `_main_modified` has been removed.  Good. Are there any consequences to this?  It also looks like `_original_main` is conditionally added. Can you consider which of these are vital to you strategy of session saving, and if there are any essential attributes that are currently dynamic, then you can add them to `__init__`.

Reply:
If I'm not wrong, `_file` was an attribute added to `Unpickler`. It was substituted in the other branch by attributes like `_file_tell` and `_file_seek`, following the `piclke.py` style with `_file_read` and `_file_write`. Here, `_main_modified` is gone and `_original_main` is conditionally set, as you noted. The logic at the relevant functions was adapted. I'll add the attributes that are no there yet to the `__init__`s.